panic: runtime error: index out of range [0] with length 0

### Relevant telegraf.conf

```toml
[global_tags]

[agent]
  # debug = true
  # quiet = false
  # logtarget = "stderr"

  interval = "5m"
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  collection_jitter = "0s"
  flush_interval = "5m"
  flush_jitter = "0s"
  precision = "0s"
  hostname = ""
  omit_hostname = false
```


### Logs from Telegraf

```text
2024-12-28T00:45:35.919455+00:00 orangepizeroplus telegraf[2083]: panic: runtime error: index out of range [0] with length 0
2024-12-28T00:45:35.926049+00:00 orangepizeroplus telegraf[2083]: goroutine 705 [running]:
2024-12-28T00:45:35.940842+00:00 orangepizeroplus telegraf[2083]: github.com/influxdata/telegraf/plugins/outputs/influxdb_v2.(*httpClient).Write(0x4002581180, {0x937c360, 0xf687960}, {0x40001e0c88, 0x25, 0x25})
2024-12-28T00:45:35.941386+00:00 orangepizeroplus telegraf[2083]: #011/go/src/github.com/influxdata/telegraf/plugins/outputs/influxdb_v2/http.go:213 +0x844
2024-12-28T00:45:35.941633+00:00 orangepizeroplus telegraf[2083]: github.com/influxdata/telegraf/plugins/outputs/influxdb_v2.(*InfluxDB).Write(0x40029e5688, {0x40001e0c88, 0x25, 0x25})
2024-12-28T00:45:35.941794+00:00 orangepizeroplus telegraf[2083]: #011/go/src/github.com/influxdata/telegraf/plugins/outputs/influxdb_v2/influxdb_v2.go:201 +0x10c
2024-12-28T00:45:35.965380+00:00 orangepizeroplus telegraf[2083]: github.com/influxdata/telegraf/models.(*RunningOutput).writeMetrics(0x40029d24d0, {0x40001e0c88, 0x25, 0x25})
2024-12-28T00:45:35.965849+00:00 orangepizeroplus telegraf[2083]: #011/go/src/github.com/influxdata/telegraf/models/running_output.go:355 +0xf8
2024-12-28T00:45:35.966121+00:00 orangepizeroplus telegraf[2083]: github.com/influxdata/telegraf/models.(*RunningOutput).Write(0x40029d24d0)
2024-12-28T00:45:35.966301+00:00 orangepizeroplus telegraf[2083]: #011/go/src/github.com/influxdata/telegraf/models/running_output.go:313 +0x38c
2024-12-28T00:45:35.966519+00:00 orangepizeroplus telegraf[2083]: github.com/influxdata/telegraf/agent.(*Agent).flushOnce.func1()
2024-12-28T00:45:35.966720+00:00 orangepizeroplus telegraf[2083]: #011/go/src/github.com/influxdata/telegraf/agent/agent.go:938 +0x30
2024-12-28T00:45:35.966897+00:00 orangepizeroplus telegraf[2083]: created by github.com/influxdata/telegraf/agent.(*Agent).flushOnce in goroutine 74
2024-12-28T00:45:35.967292+00:00 orangepizeroplus telegraf[2083]: #011/go/src/github.com/influxdata/telegraf/agent/agent.go:937 +0x94
2024-12-28T00:45:36.049336+00:00 orangepizeroplus systemd[1]: telegraf.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
2024-12-28T00:45:36.055622+00:00 orangepizeroplus systemd[1]: telegraf.service: Failed with result 'exit-code'.
2024-12-28T00:45:36.114815+00:00 orangepizeroplus systemd[1]: telegraf.service: Consumed 5.217s CPU time.
2024-12-28T00:45:36.436469+00:00 orangepizeroplus systemd[1]: telegraf.service: Scheduled restart job, restart counter is at 1.
2024-12-28T00:45:36.438362+00:00 orangepizeroplus systemd[1]: Stopped telegraf.service - Telegraf.
2024-12-28T00:45:36.439013+00:00 orangepizeroplus systemd[1]: telegraf.service: Consumed 5.217s CPU time.
2024-12-28T00:45:36.478791+00:00 orangepizeroplus systemd[1]: Starting telegraf.service - Telegraf...
```


### System info

Telegraf 1.33.0

### Docker

_No response_

### Steps to reproduce

I don't know

### Expected behavior

don't panic

### Actual behavior

is panicking

### Additional info

_No response_

I was able to reproduce this using a python server script I modified for reproducing another issue.

`server.py`:
<details>

```
#!/usr/bin/env python3

import getopt
from http.server import BaseHTTPRequestHandler, HTTPServer
import random
import sys
import time

class ErrorServerHandler(BaseHTTPRequestHandler):
    status: int = 200
    def fake_request(self):
        self.send_response(self.status)
        if self.status == 200:
            self.send_header("Content-type", "application/json")
        self.end_headers()
        if self.status == 200:
            try:
                self.wfile.write(b"{}\n")
            except Exception:
                print("%s: broken pipe (client disconnect?)" % self.path)

    def do_GET(self):
        self.fake_request()

    def do_POST(self):
        self.fake_request()

def run(port: int, status: int = 200):
    # XXX: validate delay and port
    addr = ('', port)
    httpd = HTTPServer(addr, ErrorServerHandler)
    ErrorServerHandler.status = status
    httpd.serve_forever()

if __name__ == '__main__':
    global delay
    try:
        opts, rest = getopt.getopt(sys.argv[1:], "hp:s:", [])
    except Exception:
        print("Bad args. See -h")
        sys.exit(1)

    port: int = 8000
    status: int = 200
    for opt, arg in opts:
        if opt == "-h":
            print("""Usage: server.py ARGS
 -h     this message
 -p N   listen on port N
 -s N   return status code N

GET and POST are supported an all paths and simply return an empty JSON
document.
""")
            sys.exit(0)
        elif opt == "-p":
            port = int(arg)
        elif opt == "-s":
            status = int(arg)
    try:
        run(port, status)
    except KeyboardInterrupt:
        pass
```

</details>

To reproduce:
1. start the server:
```
$ python3 ./server.py -s 400
127.0.0.1 - - [09/Jan/2025 16:01:16] "POST /api/v2/write?bucket=Telegraf_bucket_name&org=myorganization HTTP/1.1" 400 -
```
2. use this `telegraf.conf`:
```
$ cat ./telegraf.conf 
[global_tags]
[agent]
  interval = "10s"
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  collection_jitter = "0s"
  flush_interval = "10s"
  flush_jitter = "0s"
  precision = "0s"
  hostname = ""
  omit_hostname = false

[[inputs.cpu]]
  percpu = true
  totalcpu = true
  collect_cpu_time = false
  report_active = false
  [inputs.cpu.tags]
    destination = "Telegraf_bucket_name"

[[outputs.influxdb_v2]]
  urls = ["http://127.0.0.1:8000"]
  token = "mytoken"
  organization = "myorganization"
  bucket = "Telegraf"
  bucket_tag = "destination"
  exclude_bucket_tag = true
```
3. start telegraf:
```
$ telegraf --debug --config ./telegraf.conf 
2025-01-09T22:00:46Z I! Loading config: ./telegraf.conf
2025-01-09T22:00:46Z I! Starting Telegraf 1.33.0 brought to you by InfluxData the makers of InfluxDB
2025-01-09T22:00:46Z I! Available plugins: 236 inputs, 9 aggregators, 33 processors, 26 parsers, 63 outputs, 6 secret-stores
2025-01-09T22:00:46Z I! Loaded inputs: cpu
2025-01-09T22:00:46Z I! Loaded aggregators:
2025-01-09T22:00:46Z I! Loaded processors:
2025-01-09T22:00:46Z I! Loaded secretstores:
2025-01-09T22:00:46Z I! Loaded outputs: influxdb_v2
2025-01-09T22:00:46Z I! Tags enabled: host=tg16388
2025-01-09T22:00:46Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"tg16388", Flush Interval:10s
2025-01-09T22:00:46Z D! [agent] Initializing plugins
2025-01-09T22:00:46Z D! [agent] Connecting outputs
2025-01-09T22:00:46Z D! [agent] Attempting connection to [outputs.influxdb_v2]
2025-01-09T22:00:46Z D! [agent] Successfully connected to outputs.influxdb_v2
2025-01-09T22:00:46Z D! [agent] Starting service inputs
2025-01-09T22:00:56Z D! [outputs.influxdb_v2] Buffer fullness: 0 / 10000 metrics
2025-01-09T22:01:06Z D! [outputs.influxdb_v2] Buffer fullness: 21 / 10000 metrics
2025-01-09T22:01:06Z E! [agent] Error writing to outputs.influxdb_v2: Post "http://127.0.0.1:8000/api/v2/write?bucket=Telegraf_bucket_name&org=myorganization": dial tcp 127.0.0.1:8000: connect: connection refused
panic: runtime error: index out of range [0] with length 0

goroutine 103 [running]:
github.com/influxdata/telegraf/plugins/outputs/influxdb_v2.(*httpClient).Write(0xc0023d6280, {0xa461d70, 0x109d2ac0}, {0xc00243c2c8, 0x2a, 0x2a})
	/go/src/github.com/influxdata/telegraf/plugins/outputs/influxdb_v2/http.go:216 +0xa4e
github.com/influxdata/telegraf/plugins/outputs/influxdb_v2.(*InfluxDB).Write(0xc002456248, {0xc00243c2c8, 0x2a, 0x2a})
	/go/src/github.com/influxdata/telegraf/plugins/outputs/influxdb_v2/influxdb_v2.go:201 +0x131
github.com/influxdata/telegraf/models.(*RunningOutput).writeMetrics(0xc00243ac60, {0xc00243c2c8, 0x2a, 0x2a})
	/go/src/github.com/influxdata/telegraf/models/running_output.go:355 +0x10f
github.com/influxdata/telegraf/models.(*RunningOutput).Write(0xc00243ac60)
	/go/src/github.com/influxdata/telegraf/models/running_output.go:313 +0x38a
github.com/influxdata/telegraf/agent.(*Agent).flushOnce.func1()
	/go/src/github.com/influxdata/telegraf/agent/agent.go:938 +0x23
created by github.com/influxdata/telegraf/agent.(*Agent).flushOnce in goroutine 86
	/go/src/github.com/influxdata/telegraf/agent/agent.go:937 +0xa6
```

Using the telegraf binary from PR #16388, it no longer crashes:
```
$ telegraf --debug --config ./telegraf.conf 
2025-01-09T22:03:40Z I! Loading config: ./telegraf.conf
2025-01-09T22:03:40Z I! Starting Telegraf 1.34.0-1304dd90 brought to you by InfluxData the makers of InfluxDB
2025-01-09T22:03:40Z I! Available plugins: 236 inputs, 9 aggregators, 33 processors, 26 parsers, 63 outputs, 6 secret-stores
2025-01-09T22:03:40Z I! Loaded inputs: cpu
2025-01-09T22:03:40Z I! Loaded aggregators:
2025-01-09T22:03:40Z I! Loaded processors:
2025-01-09T22:03:40Z I! Loaded secretstores:
2025-01-09T22:03:40Z I! Loaded outputs: influxdb_v2
2025-01-09T22:03:40Z I! Tags enabled: host=tg16388
2025-01-09T22:03:40Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"tg16388", Flush Interval:10s
2025-01-09T22:03:40Z W! [agent] The default value of 'skip_processors_after_aggregators' will change to 'true' with Telegraf v1.40.0! If you need the current default behavior, please explicitly set the option to 'false'!
2025-01-09T22:03:40Z D! [agent] Initializing plugins
2025-01-09T22:03:40Z D! [agent] Connecting outputs
2025-01-09T22:03:40Z D! [agent] Attempting connection to [outputs.influxdb_v2]
2025-01-09T22:03:40Z D! [agent] Successfully connected to outputs.influxdb_v2
2025-01-09T22:03:40Z D! [agent] Starting service inputs
2025-01-09T22:03:50Z D! [outputs.influxdb_v2] Buffer fullness: 0 / 10000 metrics
2025-01-09T22:04:00Z E! [outputs.influxdb_v2] When writing to [http://127.0.0.1:8000/api/v2/write]: failed to write metric to Telegraf_bucket_name (will be dropped: 400 Bad Request)
2025-01-09T22:04:00Z D! [outputs.influxdb_v2] Buffer fullness: 0 / 10000 metrics
2025-01-09T22:04:00Z E! [agent] Error writing to outputs.influxdb_v2: failed to write metric to Telegraf_bucket_name (will be dropped: 400 Bad Request)
2025-01-09T22:04:10Z E! [outputs.influxdb_v2] When writing to [http://127.0.0.1:8000/api/v2/write]: failed to write metric to Telegraf_bucket_name (will be dropped: 400 Bad Request)
2025-01-09T22:04:10Z D! [outputs.influxdb_v2] Buffer fullness: 0 / 10000 metrics
2025-01-09T22:04:10Z E! [agent] Error writing to outputs.influxdb_v2: failed to write metric to Telegraf_bucket_name (will be dropped: 400 Bad Request)
...
```

--- a/plugins/outputs/influxdb_v2/http.go
+++ b/plugins/outputs/influxdb_v2/http.go
@@ -26,16 +26,17 @@ import (
 )
 
 type APIError struct {
-	StatusCode  int
-	Title       string
-	Description string
+	Err        error
+	StatusCode int
+	Retryable  bool
 }
 
 func (e APIError) Error() string {
-	if e.Description != "" {
-		return fmt.Sprintf("%s: %s", e.Title, e.Description)
-	}
-	return e.Title
+	return e.Err.Error()
+}
+
+func (e APIError) Unwrap() error {
+	return e.Err
 }
 
 const (
@@ -185,7 +186,7 @@ func (c *httpClient) Write(ctx context.Context, metrics []telegraf.Metric) error
 			}
 
 			batches[bucket] = append(batches[bucket], metric)
-			batchIndices[c.bucket] = append(batchIndices[c.bucket], i)
+			batchIndices[bucket] = append(batchIndices[bucket], i)
 		}
 	}
 
@@ -201,10 +202,14 @@ func (c *httpClient) Write(ctx context.Context, metrics []telegraf.Metric) error
 		var apiErr *APIError
 		if errors.As(err, &apiErr) {
 			if apiErr.StatusCode == http.StatusRequestEntityTooLarge {
+				// TODO: Need a testcase to verify rejected metrics are not retried...
 				return c.splitAndWriteBatch(ctx, c.bucket, metrics)
 			}
 			wErr.Err = err
-			wErr.MetricsReject = append(wErr.MetricsReject, batchIndices[bucket]...)
+			if !apiErr.Retryable {
+				wErr.MetricsReject = append(wErr.MetricsReject, batchIndices[bucket]...)
+			}
+			// TODO: Clarify if we should continue here to try the remaining buckets?
 			return &wErr
 		}
 
@@ -301,11 +306,10 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 	}
 
 	// We got an error and now try to decode further
+	var desc string
 	writeResp := &genericRespError{}
-	err = json.NewDecoder(resp.Body).Decode(writeResp)
-	desc := writeResp.Error()
-	if err != nil {
-		desc = resp.Status
+	if json.NewDecoder(resp.Body).Decode(writeResp) == nil {
+		desc = writeResp.Error()
 	}
 
 	switch resp.StatusCode {
@@ -313,9 +317,8 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 	case http.StatusRequestEntityTooLarge:
 		c.log.Errorf("Failed to write metric to %s, request was too large (413)", bucket)
 		return &APIError{
-			StatusCode:  resp.StatusCode,
-			Title:       resp.Status,
-			Description: desc,
+			Err:        fmt.Errorf("%s: %s", resp.Status, desc),
+			StatusCode: resp.StatusCode,
 		}
 	case
 		// request was malformed:
@@ -325,14 +328,10 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 		http.StatusUnprocessableEntity,
 		http.StatusNotAcceptable:
 
-		// Clients should *not* repeat the request and the metrics should be dropped.
-		rejected := make([]int, 0, len(metrics))
-		for i := range len(metrics) {
-			rejected = append(rejected, i)
-		}
-		return &internal.PartialWriteError{
-			Err:           fmt.Errorf("failed to write metric to %s (will be dropped: %s): %s", bucket, resp.Status, desc),
-			MetricsReject: rejected,
+		// Clients should *not* repeat the request and the metrics should be rejected.
+		return &APIError{
+			Err:        fmt.Errorf("failed to write metric to %s (will be dropped: %s): %s", bucket, resp.Status, desc),
+			StatusCode: resp.StatusCode,
 		}
 	case http.StatusUnauthorized, http.StatusForbidden:
 		return fmt.Errorf("failed to write metric to %s (%s): %s", bucket, resp.Status, desc)
@@ -351,13 +350,9 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 	// if it's any other 4xx code, the client should not retry as it's the client's mistake.
 	// retrying will not make the request magically work.
 	if len(resp.Status) > 0 && resp.Status[0] == '4' {
-		rejected := make([]int, 0, len(metrics))
-		for i := range len(metrics) {
-			rejected = append(rejected, i)
-		}
-		return &internal.PartialWriteError{
-			Err:           fmt.Errorf("failed to write metric to %s (will be dropped: %s): %s", bucket, resp.Status, desc),
-			MetricsReject: rejected,
+		return &APIError{
+			Err:        fmt.Errorf("failed to write metric to %s (will be dropped: %s): %s", bucket, resp.Status, desc),
+			StatusCode: resp.StatusCode,
 		}
 	}
 
@@ -367,10 +362,14 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 		desc = fmt.Sprintf("%s; %s", desc, xErr)
 	}
 
+	if desc != "" {
+		desc = ": " + desc
+	}
+
 	return &APIError{
-		StatusCode:  resp.StatusCode,
-		Title:       resp.Status,
-		Description: desc,
+		Err:        fmt.Errorf("failed to write metric to bucket %q: %s%s", bucket, resp.Status, desc),
+		StatusCode: resp.StatusCode,
+		Retryable:  true,
 	}
 }
 

--- a/plugins/outputs/influxdb_v2/http.go
+++ b/plugins/outputs/influxdb_v2/http.go
@@ -309,7 +309,7 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 	var desc string
 	writeResp := &genericRespError{}
 	if json.NewDecoder(resp.Body).Decode(writeResp) == nil {
-		desc = writeResp.Error()
+		desc = ": " + writeResp.Error()
 	}
 
 	switch resp.StatusCode {
@@ -330,11 +330,11 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 
 		// Clients should *not* repeat the request and the metrics should be rejected.
 		return &APIError{
-			Err:        fmt.Errorf("failed to write metric to %s (will be dropped: %s): %s", bucket, resp.Status, desc),
+			Err:        fmt.Errorf("failed to write metric to %s (will be dropped: %s)%s", bucket, resp.Status, desc),
 			StatusCode: resp.StatusCode,
 		}
 	case http.StatusUnauthorized, http.StatusForbidden:
-		return fmt.Errorf("failed to write metric to %s (%s): %s", bucket, resp.Status, desc)
+		return fmt.Errorf("failed to write metric to %s (%s)%s", bucket, resp.Status, desc)
 	case http.StatusTooManyRequests,
 		http.StatusServiceUnavailable,
 		http.StatusBadGateway,
@@ -351,7 +351,7 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 	// retrying will not make the request magically work.
 	if len(resp.Status) > 0 && resp.Status[0] == '4' {
 		return &APIError{
-			Err:        fmt.Errorf("failed to write metric to %s (will be dropped: %s): %s", bucket, resp.Status, desc),
+			Err:        fmt.Errorf("failed to write metric to %s (will be dropped: %s)%s", bucket, resp.Status, desc),
 			StatusCode: resp.StatusCode,
 		}
 	}
@@ -359,11 +359,7 @@ func (c *httpClient) writeBatch(ctx context.Context, bucket string, metrics []te
 	// This is only until platform spec is fully implemented. As of the
 	// time of writing, there is no error body returned.
 	if xErr := resp.Header.Get("X-Influx-Error"); xErr != "" {
-		desc = fmt.Sprintf("%s; %s", desc, xErr)
-	}
-
-	if desc != "" {
-		desc = ": " + desc
+		desc = fmt.Sprintf(": %s; %s", desc, xErr)
 	}
 
 	return &APIError{

--- a/plugins/outputs/influxdb_v2/influxdb_v2.go
+++ b/plugins/outputs/influxdb_v2/influxdb_v2.go
@@ -199,11 +199,11 @@ func (i *InfluxDB) Write(metrics []telegraf.Metric) error {
 	for _, n := range rand.Perm(len(i.clients)) {
 		client := i.clients[n]
 		if err := client.Write(ctx, metrics); err != nil {
+			i.Log.Errorf("When writing to [%s]: %v", client.url, err)
 			var werr *internal.PartialWriteError
 			if errors.As(err, &werr) || errors.Is(err, internal.ErrSizeLimitReached) {
 				return err
 			}
-			i.Log.Errorf("When writing to [%s]: %v", client.url, err)
 			continue
 		}
 		return nil

--- a/plugins/outputs/influxdb_v2/influxdb_v2_test.go
+++ b/plugins/outputs/influxdb_v2/influxdb_v2_test.go
@@ -1,11 +1,13 @@
 package influxdb_v2_test
 
 import (
+	"fmt"
 	"io"
 	"net"
 	"net/http"
 	"net/http/httptest"
 	"reflect"
+	"strconv"
 	"strings"
 	"sync/atomic"
 	"testing"
@@ -485,4 +487,375 @@ func TestRateLimit(t *testing.T) {
 	time.Sleep(time.Duration(plugin.RateLimitConfig.Period) + 100*time.Millisecond)
 	require.NoError(t, plugin.Write(metrics[3:]))
 	require.Equal(t, uint64(121), received.Load())
+}
+
+func TestStatusCodeNonRetryable4xx(t *testing.T) {
+	codes := []int{
+		// Explicitly checked non-retryable status codes
+		http.StatusBadRequest, http.StatusUnprocessableEntity, http.StatusNotAcceptable,
+		// Other non-retryable 4xx status codes not explicitly checked
+		http.StatusNotFound, http.StatusExpectationFailed,
+	}
+
+	for _, code := range codes {
+		t.Run(fmt.Sprintf("code %d", code), func(t *testing.T) {
+			// Setup a test server
+			ts := httptest.NewServer(
+				http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					body, err := io.ReadAll(r.Body)
+					if err != nil {
+						t.Error(err)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+					if strings.Contains(string(body), "bucket=foo") {
+						w.WriteHeader(http.StatusOK)
+						return
+					}
+					w.WriteHeader(code)
+				}),
+			)
+			defer ts.Close()
+
+			// Setup plugin and connect
+			plugin := &influxdb.InfluxDB{
+				URLs:      []string{"http://" + ts.Listener.Addr().String()},
+				BucketTag: "bucket",
+				Log:       &testutil.Logger{},
+			}
+			require.NoError(t, plugin.Init())
+			require.NoError(t, plugin.Connect())
+			defer plugin.Close()
+
+			// Together the metric batch size is too big, split up, we get success
+			metrics := []telegraf.Metric{
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "foo",
+					},
+					map[string]interface{}{
+						"value": 0.0,
+					},
+					time.Unix(0, 0),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "my_bucket",
+					},
+					map[string]interface{}{
+						"value": 42.0,
+					},
+					time.Unix(0, 1),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "my_bucket",
+					},
+					map[string]interface{}{
+						"value": 43.0,
+					},
+					time.Unix(0, 2),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "foo",
+					},
+					map[string]interface{}{
+						"value": 0.0,
+					},
+					time.Unix(0, 3),
+				),
+			}
+
+			// Write the metrics the first time and check for the expected errors
+			err := plugin.Write(metrics)
+			require.ErrorContains(t, err, "failed to write metric to my_bucket (will be dropped:")
+
+			var apiErr *influxdb.APIError
+			require.ErrorAs(t, err, &apiErr)
+			require.Equal(t, code, apiErr.StatusCode)
+
+			var writeErr *internal.PartialWriteError
+			require.ErrorAs(t, err, &writeErr)
+			require.Len(t, writeErr.MetricsReject, 2, "rejected metrics")
+		})
+	}
+}
+
+func TestStatusCodeInvalidAuthentication(t *testing.T) {
+	codes := []int{http.StatusUnauthorized, http.StatusForbidden}
+
+	for _, code := range codes {
+		t.Run(fmt.Sprintf("code %d", code), func(t *testing.T) {
+			// Setup a test server
+			ts := httptest.NewServer(
+				http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					body, err := io.ReadAll(r.Body)
+					if err != nil {
+						t.Error(err)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+					if strings.Contains(string(body), "bucket=foo") {
+						w.WriteHeader(http.StatusOK)
+						return
+					}
+					w.WriteHeader(code)
+				}),
+			)
+			defer ts.Close()
+
+			// Setup plugin and connect
+			plugin := &influxdb.InfluxDB{
+				URLs:      []string{"http://" + ts.Listener.Addr().String()},
+				BucketTag: "bucket",
+				Log:       &testutil.Logger{},
+			}
+			require.NoError(t, plugin.Init())
+			require.NoError(t, plugin.Connect())
+			defer plugin.Close()
+
+			// Together the metric batch size is too big, split up, we get success
+			metrics := []telegraf.Metric{
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "foo",
+					},
+					map[string]interface{}{
+						"value": 0.0,
+					},
+					time.Unix(0, 0),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "my_bucket",
+					},
+					map[string]interface{}{
+						"value": 42.0,
+					},
+					time.Unix(0, 1),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "my_bucket",
+					},
+					map[string]interface{}{
+						"value": 43.0,
+					},
+					time.Unix(0, 2),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "foo",
+					},
+					map[string]interface{}{
+						"value": 0.0,
+					},
+					time.Unix(0, 3),
+				),
+			}
+
+			// Write the metrics the first time and check for the expected errors
+			err := plugin.Write(metrics)
+			require.ErrorContains(t, err, "failed to write metric to my_bucket")
+			require.ErrorContains(t, err, strconv.Itoa(code))
+
+			var writeErr *internal.PartialWriteError
+			require.ErrorAs(t, err, &writeErr)
+			require.Empty(t, writeErr.MetricsReject, "rejected metrics")
+			require.LessOrEqual(t, len(writeErr.MetricsAccept), 2, "accepted metrics")
+		})
+	}
+}
+
+func TestStatusCodeServiceUnavailable(t *testing.T) {
+	codes := []int{
+		http.StatusTooManyRequests,
+		http.StatusServiceUnavailable,
+		http.StatusBadGateway,
+		http.StatusGatewayTimeout,
+	}
+
+	for _, code := range codes {
+		t.Run(fmt.Sprintf("code %d", code), func(t *testing.T) {
+			// Setup a test server
+			ts := httptest.NewServer(
+				http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					body, err := io.ReadAll(r.Body)
+					if err != nil {
+						t.Error(err)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+					if strings.Contains(string(body), "bucket=foo") {
+						w.WriteHeader(http.StatusOK)
+						return
+					}
+					w.WriteHeader(code)
+				}),
+			)
+			defer ts.Close()
+
+			// Setup plugin and connect
+			plugin := &influxdb.InfluxDB{
+				URLs:      []string{"http://" + ts.Listener.Addr().String()},
+				BucketTag: "bucket",
+				Log:       &testutil.Logger{},
+			}
+			require.NoError(t, plugin.Init())
+			require.NoError(t, plugin.Connect())
+			defer plugin.Close()
+
+			// Together the metric batch size is too big, split up, we get success
+			metrics := []telegraf.Metric{
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "foo",
+					},
+					map[string]interface{}{
+						"value": 0.0,
+					},
+					time.Unix(0, 0),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "my_bucket",
+					},
+					map[string]interface{}{
+						"value": 42.0,
+					},
+					time.Unix(0, 1),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "my_bucket",
+					},
+					map[string]interface{}{
+						"value": 43.0,
+					},
+					time.Unix(0, 2),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "foo",
+					},
+					map[string]interface{}{
+						"value": 0.0,
+					},
+					time.Unix(0, 3),
+				),
+			}
+
+			// Write the metrics the first time and check for the expected errors
+			err := plugin.Write(metrics)
+			require.ErrorContains(t, err, "waiting 25ms for server (my_bucket) before sending metric again")
+
+			var writeErr *internal.PartialWriteError
+			require.ErrorAs(t, err, &writeErr)
+			require.Empty(t, writeErr.MetricsReject, "rejected metrics")
+			require.LessOrEqual(t, len(writeErr.MetricsAccept), 2, "accepted metrics")
+		})
+	}
+}
+
+func TestStatusCodeUnexpected(t *testing.T) {
+	codes := []int{http.StatusInternalServerError}
+
+	for _, code := range codes {
+		t.Run(fmt.Sprintf("code %d", code), func(t *testing.T) {
+			// Setup a test server
+			ts := httptest.NewServer(
+				http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					body, err := io.ReadAll(r.Body)
+					if err != nil {
+						t.Error(err)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+					if strings.Contains(string(body), "bucket=foo") {
+						w.WriteHeader(http.StatusOK)
+						return
+					}
+					w.WriteHeader(code)
+				}),
+			)
+			defer ts.Close()
+
+			// Setup plugin and connect
+			plugin := &influxdb.InfluxDB{
+				URLs:      []string{"http://" + ts.Listener.Addr().String()},
+				BucketTag: "bucket",
+				Log:       &testutil.Logger{},
+			}
+			require.NoError(t, plugin.Init())
+			require.NoError(t, plugin.Connect())
+			defer plugin.Close()
+
+			// Together the metric batch size is too big, split up, we get success
+			metrics := []telegraf.Metric{
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "foo",
+					},
+					map[string]interface{}{
+						"value": 0.0,
+					},
+					time.Unix(0, 0),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "my_bucket",
+					},
+					map[string]interface{}{
+						"value": 42.0,
+					},
+					time.Unix(0, 1),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "my_bucket",
+					},
+					map[string]interface{}{
+						"value": 43.0,
+					},
+					time.Unix(0, 2),
+				),
+				metric.New(
+					"cpu",
+					map[string]string{
+						"bucket": "foo",
+					},
+					map[string]interface{}{
+						"value": 0.0,
+					},
+					time.Unix(0, 3),
+				),
+			}
+
+			// Write the metrics the first time and check for the expected errors
+			err := plugin.Write(metrics)
+			require.ErrorContains(t, err, "failed to write metric to bucket \"my_bucket\"")
+			require.ErrorContains(t, err, strconv.Itoa(code))
+
+			var writeErr *internal.PartialWriteError
+			require.ErrorAs(t, err, &writeErr)
+			require.Empty(t, writeErr.MetricsReject, "rejected metrics")
+			require.LessOrEqual(t, len(writeErr.MetricsAccept), 2, "accepted metrics")
+		})
+	}
 }

